Issue 18 stream read

dadlex opened the pull request.

dadlex committed:
Issue-19-StreamRead

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Add invokeStreamRead method which takes a PathOperation input parameter, parses its dstPath field and reads all the available events. It does not store them anywhere, just prints out. Timeout is set to 100, which is a bit random, values like 0 or 1 work awfully.
Checks if the scope or stream exist are made in submit switch method

Signed-off-by: Alexander Aphonin<alex15051997@yandex.com>

dadlex committed:
Delete a line with comments

delete a meaningless line

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -177,6 +186,48 @@ protected int submit(List<O> ops) throws InterruptRunException, IllegalStateExce
 
+    
+	private boolean invokePathRead(final PathOperation<? extends PathItem> pathOp) {
+		final int READER_TIMEOUT_MS = 100;
```
1. Why not 1 instead of 100?
2. Why not to place in some kind of interface like "PravegaConstants"?

dadlex replied:
1. because TIMEOUT has some sort of a bug. And its value must be big enough to start reading the stream. It contradicts with docs, but i couldn't get consistent results with 1.
2. I've moved it to PravegaConstants.java file. Should it be an interface as you said?

akurilov replied:
There are constants interface already existing:
https://github.com/emc-mongoose/mongoose-storage-driver-pravega/blob/master/src/main/java/com/emc/mongoose/storage/driver/pravega/Constants.java
Please update your branch first

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -177,6 +186,48 @@ protected int submit(List<O> ops) throws InterruptRunException, IllegalStateExce
+    
+	private boolean invokePathRead(final PathOperation<? extends PathItem> pathOp) {
+		final int READER_TIMEOUT_MS = 100;
+		StreamManager streamManager = StreamManager.create(controllerURI);
```
It may be useful to use "final" where possible

evtimo replied:
Do we need to declare StreamManager here or in PravegaStorageDriver class (48 line) ?
Or we have a new object of StreamManager and ClientFactory on every call?

akurilov replied:
Good question. Ideally we should reuse the stream manager instance in the same way as scopes and streams. E.g. we should have a concurrent dictionary of the stream managers mapped by the controller URI as keys. I thought that we will do this a bit later and did not requested right now. First we should make it just working and later we can refactor this for the optimal performance.

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -177,6 +186,48 @@ protected int submit(List<O> ops) throws InterruptRunException, IllegalStateExce
+		final ReaderGroupConfig readerGroupConfig = ReaderGroupConfig.builder()
+				.stream(Stream.of(scope, streamName))
+				.build();
+		try (ReaderGroupManager readerGroupManager = ReaderGroupManager.withScope(scope, controllerURI)) {
```
It may be useful to use "final" where possible

akurilov commented:
Please add the info about the timeout to the Read section of the README file

evtimo commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -134,7 +134,16 @@ protected boolean submit(O op) throws InterruptRunException, IllegalStateExcepti
+							//putIf or computeIf
+							//probably put inside to avoid code repeat
+						};
+                        if(invokePathRead((PathOperation<? extends PathItem>) op)) {
```
Why not to use pathOp, declared in line 126?

dadlex replied:
Changed. Agree. But to avoid an unchecked assignment then we need to cast like this:
`final PathOperation<? extends PathItem> pathOp = (PathOperation<? extends PathItem>) op;`
in else branch for PathItem.

evtimo commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -134,7 +134,16 @@ protected boolean submit(O op) throws InterruptRunException, IllegalStateExcepti
+                    	final String path = pathOp.dstPath();
+						final String scopeName = path.substring(0,path.indexOf("/"));
+						final String streamName = path.substring(path.indexOf("/")+1, path.length());
+						if (scopeMap.get(scopeName).get(streamName)==null) {
```
Do we need to check this on READ operation? If we can't read from this path should we to create it? To my mind, it is applicable only for CREATE/UPDATE operations and throw exception/error/NULL in case we can't read.

akurilov replied:
You're right. We should not create any scope/stream if we are trying to read. The app should set the corresponing status ("not found") and log error message in that case

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -134,7 +134,16 @@ protected boolean submit(O op) throws InterruptRunException, IllegalStateExcepti
+                    	final String path = pathOp.dstPath();
+						final String scopeName = path.substring(0,path.indexOf("/"));
+						final String streamName = path.substring(path.indexOf("/")+1, path.length());
+						if (scopeMap.get(scopeName).get(streamName)==null) {
```
Also, this code may throw NPE if there's no such scope

dadlex committed:
Add constant, fix scope check, add logging, add readme paragraph

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Add a constant for a default amount of segments.
Correct the check on scopes in map. It might be useless as we are about to use http check.
Add logging and mapping to internal error
Fix minor things like final and duplicated variables
Add a paragraph about timeout in readme

Signed-off-by: Alexander <alex15051997@yandex.com>

dadlex committed:
Add constant, fix scope check, add logging, add readme paragraph

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Add a constant for a default amount of segments.
Correct the check on scopes in map. It might be useless as we are about to use http check.
Add logging and mapping to internal error
Fix minor things like final and duplicated variables
Add a paragraph about timeout in readme

Signed-off-by: Alexander <alex15051997@yandex.com>

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:
```diff
@@ -172,7 +176,18 @@ protected boolean submit(O op) throws InterruptRunException, IllegalStateExcepti
+						final String path = pathOp.dstPath();
+						final String scopeName = path.substring(0, path.indexOf("/"));
+						final String streamName = path.substring(path.indexOf("/") + 1);
+						if ((scopeMap.get(scopeName) == null) || (scopeMap.get(scopeName).get(streamName) == null)) {
```
why not nested if?
```
final ... scope = scopeMap.get(scopeName);
if(null != scope) {
   final ... stream = scope.get(streamName);
   if(null != stream) {
         ...
   } else {
         ...
   }
} else {
    ...
}
```

dadlex replied:
@akurilov I would have done it if we had to handle these situations differently. I don't quite see why we need nesting if both conditions are handled in the same way. Though i will fix it for sure, i would also be happy if you give me a reason.

evtimo replied:
I think we should to replace IF condition in 182 line on 
`if ((null != scopeMap.get(scopeName) && null != scopeMap.get(scopeName).get(streamName)))`
Because now case with NULL.get(streamName) doesn't handle at all.
I also think that nested if here is unnecessary.
